### ISSUE TITLE
Reconstruct lgb using lightgbm.LGBMRegressor

### ISSUE THREAD

**Issue by joyangyi** [microsoft/LightGBM#5772]
## Description
I'm building a main classifier model (lightgbm), together with an adversary model. I tried to use lightgbm.LGBMRegressor to build tree by tree, so that I can update my main classifier model with gradient from adversary each round. 

Is lightgbm.LGBMRegressor comparable to sklearn.tree.DecisionTreeRegressor in the sense that I can reconstruct the whole boosting process manually? 

## Reproducible example
tree = lgb.LGBMRegressor()
tree.set_params(**self.params)
self.trees.append(tree)

## Environment info
Jupyter notebook

## Additional Comments
I got worse model performance on a hold out set, compared with building lgb.train directly (even if I mute adversary model). Is there anything I did wrong using LGBMRegressor to reconstruct lgb? Thanks!

**Comment by Sarah018** [comment]
Hi @jmoralez, I get same values from tree.predict(X, raw_score=False), no matter I set 'raw_score' to either True or False. The output is always raw scores. May I know why is that? I look forward to your reply.

**Comment by Sarah018** [comment]
Hi @jameslamb, could you please help me with this quesiton? I appreciate your time :)

**Comment by Sarah018** [comment]
here is an example code for testing:

**Comment by Sarah018** [comment]
import lightgbm as lgb
from sklearn.datasets import make_regression
from scipy.special import expit

X, y = make_regression()
ds = lgb.Dataset(X, y)
params = {'num_leaves': 8}
bst = lgb.Booster(params, ds)
bst.update(fobj=None) 
preds_prob = bst.predict(X, raw_score=False)  
preds_raw = bst.predict(X, raw_score=True)

**Comment by jmoralez** [comment]
The raw score argument is used when a transformation is normally applied to the raw predictions before returning the model scores, e.g. sigmoid or softmax in classification. For regression, however, the model raw score is directly used as the prediction, so if you're doing regression it's expected for these to be the same.

**Comment by Sarah018** [comment]
Hi @jmoralez, sorry I forgot to include the definition of gradient, here it is: gradient = y - 1/(1+np.exp(-logit))

**Comment by Sarah018** [comment]
I'm putting together the whole block of code for you to review, I hope it's easier for you to read:

params ={…, 'objective': 'regression',  'n_estimators':1}

logit = np.log(np.sum(y)/np.sum(1-y))

for i in range(max_iter):
        gradient = y - 1/(1+np.exp(-logit))
	ds = lightgbm.Dataset(X, gradient)
	tree = lightgbm.Booster(self.params, ds)
	tree.update(fobj=None)  
	update = tree.predict(X)  # predictions (at log odds scale) after a single iteration
	logit += np.multiply(self.learning_rate, update) 
	y_fin = 1/(1+np.exp(-logit)) # sigmoid of logit. Probabiltity of an event. Not involved in tree updating.

**Comment by Sarah018** [comment]
Hi @jmoralez, I put together the whole block of code for you to review, I hope it's easier for you to read:

params ={…, 'objective': 'regression', 'n_estimators':1}

logit = np.log(np.sum(y)/np.sum(1-y))

for i in range(max_iter):
        gradient = y - 1/(1+np.exp(-logit))
        ds = lightgbm.Dataset(X, gradient)
        tree = lightgbm.Booster(self.params, ds)
        tree.update(fobj=None)
        update = tree.predict(X) # predictions (at log odds scale) after a single iteration
        logit += np.multiply(self.learning_rate, update)
        y_fin = 1/(1+np.exp(-logit)) # sigmoid of logit. Probabiltity of an event. Not involved in tree updating.

**Comment by Sarah018** [comment]
Hi @jmoralez, I put together the whole block of code for you to review, I hope it's easier for you to read:

params ={…, 'objective': 'regression', 'n_estimators':1}

logit = np.log(np.sum(y)/np.sum(1-y))

for i in range(max_iter):
        gradient = y - 1/(1+np.exp(-logit))
        ds = lightgbm.Dataset(X, gradient)
        tree = lightgbm.Booster(self.params, ds)
        tree.update(fobj=None)
        update = tree.predict(X) # predictions (at log odds scale) after a single iteration
        logit += np.multiply(self.learning_rate, update)
        y_fin = 1/(1+np.exp(-logit)) # sigmoid of logit. Probabiltity of an event. Not involved in tree updating.

**Comment by jmoralez** [comment]
Hi. Seems like you may be better off using `fobj` instead (see the note at the bottom [here](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train) for reference), since that allows you to manually compute the gradients and hessians used in the boosting process.

**Comment by Sarah018** [comment]
Hi @jmoralez, thanks a lot for your valuable suggestions, I really appreciate it! I will give them a try 👍

**Comment by Sarah018** [comment]
Hi @jmoralez,

Thank you so much for your helpful suggestions on manually building LGBM tree by tree! Your advice truly made a big difference in improving our methodology and achieving better results. However, we've run into a bit of a snag and we were hoping you might be able to help us out.

To clarify, we attempted to reproduce the model using two different approaches: first by calling lgb.train, and secondly by using lgb.Booster/update to build the LGBM tree by tree.

**Our standing is that the model built by calling lgb.train should be the same as the model built by calling lgb.Booster/update as long as we use the same parameters**. Unfortunately we weren't successful.

Would you mind taking a quick look at our code? If you have any suggestions or if you see anything that doesn't look quite right, we would really appreciate your feedback. Thank you again for all of your help so far!


(1) Method A: build LGBM  tree by tree with  lgb.Booster/update

max_iter = 252

model_A = []

 

logit = np.full(np.shape(y), 0)

for i in range(max_iter):

            gradient = y - 1/(1+np.exp(-logit))

            ds = lgb.Dataset(X, gradient)

            tree = lgb.Booster(params, ds)

            tree.update(fobj=None)

            update = tree.predict(X)

            logit += np.multiply(self.learning_rate, update)

            model_A.append(tree)

 

 

(2) Method B: build LGBM  with lgb.train

model_B = lgb.train(params,

                  train_data,

                  valid_sets=valid_data,

                  num_boost_round=252,

                    keep_training_booster = True

                 )

**Comment by jmoralez** [comment]
Hi. Would you mind us taking a step back to be clear on the objective? I don't fully understand what you're trying to achieve. If you want to reproduce the behavior of `lightgbm.train` you need to keep the dataset fixed, e.g.
```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=1_000, n_features=2, n_informative=2, random_state=0)
params = {'num_leaves': 5, 'verbosity': -1}
num_rounds = 5
ds = lgb.Dataset(X, y)
# regular train
bst_train = lgb.train(params, ds, num_boost_round=num_rounds)
preds_train = bst_train.predict(X)
# manual train
bst_manual = lgb.Booster(params, ds)
for _ in range(num_rounds):
    bst_manual.update(fobj=None)
preds_manual = bst_manual.predict(X)
# compare equal predictions
np.testing.assert_allclose(preds_train, preds_manual)
```
If you want to optimize a different objective you can use `fobj` either in `Booster.update` (with the manual training) or passing it through `lightgbm.train`.

By redefining the dataset you're essentially boosting only for one iteration and the initial score is different each time (each time is the mean of the current target) and it has a higher weight than the single boosting iteration.

I think you may want to use `fobj` to set the gradients and use a constant hessian as with the L2 objective, but I can't really tell.

**Comment by johnkimdy** [comment]
Hello Jose, 
thank you for your response, it helped us a lot in our problem solving. 
We have a couple more questions.

1. If we want to build a lightgbm Booster from scratch (without custom loss function), do we have to specify the first leaf's formula or does .update() initialize it for us?
2. When we specify customized loss function and use .predict(), it only outputs raw_score. Is this true? Does .update() also always output raw_score when using custom loss func?

**Comment by jmoralez** [comment]
Hi.
1. The boosting starts from the `init_score`, which is set by the objective. If you use a custom objective it's set to 0 but you can override it by passing the argument to the [Dataset](https://lightgbm.readthedocs.io/en/v3.3.2/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset).
2. I think this is related to 1., when you use a custom objective the `init_score` is 0 and the prediction is always the raw score, so if you use a different init_score you have to add it back to your predictions.

Hope this helps.
